Add random circuit with graph

MozammilQ opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
The function `qiskit.circuit.random.utils.random_circuit_from_graph` adds a new feature
of generating random circuits with interaction graph.

### Details and comments

MozammilQ committed:
Added random_circuit_with_graph

MozammilQ committed:
added tests

MozammilQ committed:
added release note

MozammilQ committed:
lint

MozammilQ committed:
fix docs lint

MozammilQ committed:
fix lint

MozammilQ committed:
refactor code;lint

sbrandhsn commented:
Dear @MozammilQ thanks for showing interest in this issue. I will be on vacation and look into this PR next week! In the meantime make yourself familiar with https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md and the unitaryhack rules (if you have not already). :-)

MozammilQ committed:
refactor code;lint

MozammilQ committed:
refactor code;lint

MozammilQ reviewed (commented):
I have added the suggestions :)

MozammilQ commented:
@sbrandhsn , please see if this is good enough :)

sbrandhsn commented:
@MozammilQ thanks for your interest in looking at issue #12364 ! I agree this issue is a bit tricky but I think your contribution is going into the right direction. :-) Also, thanks for exploring different options for that issue!

I did not have time yet to look into your code but wanted to spend the time to clarify some of your questions. To give a little bit of context, the output of this procedure will likely not be executed directly on a quantum computer but will prove valuable for investigating compiler capabilities and efficiencies. The problem I wanted to fix with the `depth` parameter is that the given input graph may not be observed by `random_circuit_from_graph ` otherwise. Likely the name of `depth` is not optimal and should be replaced by something like `min_2q_gate_per_edge`.

You rightfully noted that the generated circuits may become quite large but I think this is fine. I suspect that if you have `N` edges in the input graph and uniformly draw `10*N` random edges from your input graph, you are pretty much guaranteed to have drawn every edge at least one. While this implies that the circuit depth of the random circuit can be `10*N` occasionally, the circuit depth should be much smaller in practice, depending on the structure of the input graph. If an user decides to assign non-uniform weights to the input graph edges, they will need to accept that the returned random circuit will be deeper.

I think it would be a good idea to raise an error with a reasonable error message if some of the edge weights are none while others are not none. If all edge weights are none, you can assume the weight of each edge to be 1/N. 

With respects to resets and conditional operations, you can simply follow the approach given in the original `random_circuit` function, i.e. meaning make some operations conditional with some probability and regard resets as single-qubit gates. 

Some of your 'reviewer markers'  should actually be comments in your code, it would be great if you add those as you see fit. :-)

I think the implementation should eventually consider the edge weight as well as consider the `depth`/`min_2q_gate_per_edge` parameter, e.g. guarantee that the returned random circuit actually has the interaction graph specified by `interaction_graph` when `depth`/`min_2q_gate_per_edge` is set to one.

I like the idea to fill unused qubits in a layer with a single-qubit gate if `insert_1q_oper` is set and otherwise do not use single-qubit gates at all. However, a layer should consist of multiple two-qubit gates (even though the number of two-qubit gates does not have be maximal in each layer). You are also right about `max_operands` - it is confusing in the way the issue is currently stated. In the future, one might want to include 3-qubit or 4-qubit gates but currently for this issue only two-qubit gates and optional single-qubit gates should be used. 

Let me know if things became clearer and whether you require any other information. Also, if you make modifications to your PR, let me know for which implementation you went for eventually. :-)

MozammilQ commented:
I have already worked a lot on this PR, so I will surely do some modifications to the code, but, actually right now bit busy in "IBM Quantum Challenge 2024"
will come back to it, after 4 days :) 

@sbrandhsn , you explained it very very well, thanks a lot for the time you gave to write and explain these.

sbrandhsn commented:
Absolutely! :-)

MozammilQ committed:
refactor code; added suggestions

MozammilQ commented:
This is python so, focus is on code readability.
Therefore, I am leaving some code repetetions deliberately :)

@sbrandhsn , please have a look, and let me know what to change.
I think reading the docstring will suffice for the explanation of the way the logic works in this PR.

MozammilQ committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ committed:
fix weird merge conflict

MozammilQ committed:
edit docstrings; fix typos

MozammilQ committed:
added a test;fixed lint

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+    If :arg:`max_operands` is set to 2, then in every iteration `N` 2Q gates and qubit-pairs
+    are chosen at random, the qubit-pairs are also chosen at random based on the probability
+    attached to the qubit-pair, the 2Q gates are applied on the qubit-paris which are unused
+    for that iteration, in case for a qubit-pair where target or control is already used up
```
What do you mean here with "the 2Q gates are applied on the qubit-pairs which are unused for that iteration"?

MozammilQ replied:
Added a better explanation :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+
+    Args:
+        interaction_graph (int): Interaction Graph
+        min_2q_gate_per_edge (int): Minimum number of times every qubit-pair must be used.
```
```suggestion
        min_2q_gate_per_edge (int): Minimum number of times every qubit-pair must be used
        in the random circuit.
```

MozammilQ replied:
Added suggestion :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+        max_operands (int): maximum qubit operands of each gate (between 1 and 2)
+        measure (bool): if True, measure all qubits at the end
+        conditional (bool): if True, insert middle measurements and conditionals
+        reset (bool): if True, insert middle resets. ( insert_1q_oper should be true )
```
I would either expect resets to be applied independently of `insert_1q_oper` or an error to be raised when `reset and not insert_1q_oper`

MozammilQ replied:
done :) 
 If `insert_1q_oper` is disabled, and `reset` enabled, then this raises `CircuitError`.
 and, a relevant message is shown to the user,  to correct that error :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+
+    Raises:
+        CircuitError: when invalid options given
+        ValueError: when the probabilities for qubit-pairs are passed and they do not add to 1.0.
```
According to the issue, the probabilities should be normalized

MozammilQ replied:
Added suggestion :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+        ValueError: when some edges have probability as None.
+    """
+    # Leaving out 'free_nodes'
+    pydi_graph, pydi_graph_node_map, _, _ = interaction_graph
```
`interaction_graph` should just be a single directed graph where vertices are qubits and edges are permitted qubit-interactions. The assignment to qubit-index to vertex can be chosen arbitrarily.

MozammilQ replied:
Did the changes serve this comment?

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+        qubits = list(pydi_graph_node_map.keys())
+    else:
+        n_q = pydi_graph.num_nodes()
+        qubits = [Qubit(QuantumRegister(n_q, "q"), idx) for idx in pydi_graph.nodes()]
```
Please create `QuantumRegister` once outside the loop.

MozammilQ replied:
Added suggestion :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+        return QuantumCircuit()
+
+    if max_operands < 1 or max_operands > 2:
+        raise CircuitError("max_operands must be between 1 and 2")
```
This function is intended to only work on single-qubit and two-qubit gates - n-qubit gates with n>2 may be supported in the future (There is a question on how to model qubit connectivity for n-qubit gates).

MozammilQ replied:
Added suggestion :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+                        qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                elif control_qubit in qubit_idx_used and target_qubit in qubit_idx_used:
+                    continue
+
```
Why couldn't you reuse the slack logic in the original `random_circuit`? I'm not sure this code is correct - what about qubits that are neither on a target/control qubit, is this possible?

MozammilQ replied:
Yes, indeed, the logic was wrong, I have changed the logic. Its explained in the docstring.

sbrandhsn commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,35 @@
+    If float values are present as probabilities but some are None, this will raise a ValueError.
+
+    Arguments accepted by the function are
+    :code:`interaction_graph` Takes in the interaction graph
```
Describe what the interaction graph is here

MozammilQ replied:
Done :)

sbrandhsn commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,35 @@
+      pydi_graph.add_nodes_from(range(n_q))
+      cp_map = [(0, 2, 0.1), (1, 3, 0.15), (2, 4, 0.15), (3, 4, 0.1), (5, 7, 0.13), (4, 7, 0.07), (7, 9, 0.1), (5, 8, 0.1), (6, 9, 0.1)]
+      pydi_graph.add_edges_from(cp_map)
+      inter_graph = (pydi_graph, None, None, None)
```
`inter_graph` should be an directed graph

MozammilQ replied:
Done :) 
The example uses, "Directed Heavy Hex Graph" with some arbitrary weights.

sbrandhsn commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -160,3 +163,196 @@ def test_random_circuit_with_zero_distribution(self):
+            seed=486,  # Do not change the seed or the args.
+            insert_1q_oper=True,
+        )
+        self.assertIn("reset", qc.count_ops())
```
You could also check for conditional gates here

MozammilQ replied:
strategy was to focus on testing only one feature in one test, but I will add code to check for conditionals here only.

MozammilQ replied:
Done :)

sbrandhsn commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -160,3 +163,196 @@ def test_random_circuit_with_zero_distribution(self):
+
+        # make sure every qubit-pair from the circuit actually present in the edge_list
+        for cp in cp_mp:
+            self.assertTrue(cp in edge_list)
```
You can replace this chunk of code by `build_interaction_graph` in https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/layout/vf2_utils.py and a subgraph isomorphism check. :-)

MozammilQ replied:
I have kept the logic untouched, because for some reasons `build_interaction_graph` is giving out a coupling map, which is not correct for the given circuit.
On visual inspection, it was confirmed that the coupling map given out by this logic is indeed correct and serve the purpose.

sbrandhsn commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -160,3 +163,196 @@ def test_random_circuit_with_zero_distribution(self):
+    def test_large_conditional_weighted_qubits(self):
+        """Test that conditions do not fail with large conditionals.  Regression test of gh-6994."""
+        # This is to test the call actually returns without raising an exception.
+        # In this case every qubit-pair is associated with a probability of being selected.
```
Please do not copy code but try to reuse it!

MozammilQ replied:
@sbrandhsn , you mean not copying and reusing code from line: 174 to line: 190 ?? ( I mean the code block under `def setUp` at line: 172 )

sbrandhsn replied:
It appears that some of the included tests have been part of the test suite for `random_circuit`, it would be nice if you could reuse these instead of copying the code. :-)

MozammilQ replied:
I have changed the tests :)

sbrandhsn reviewed (changes requested):
Thanks @MozammilQ, this already looks like it is in good shape! :-) I left a couple of comments. I also think there should be a couple of more test cases with different types of interaction graphs such as sparse graphs, (almost) fully connected and a couple more random ones. Also, I think it would be best if you reused the code in other test cases instead of copying them. See https://networkx.org/documentation/stable/reference/generators.html for how to generate these graphs.

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+    """
+    # Leaving out 'free_nodes'
+    pydi_graph, pydi_graph_node_map, _, _ = interaction_graph
+    edges_probs = pydi_graph.edges()
```
`edges_probs` may also be an empty list, your code should handle that.

MozammilQ replied:
done :)

MozammilQ replied:
L204, it just returns circuit, made by `random_circuit`

MozammilQ commented:
I am committing some unexpectedly silly mistakes here, will come back to this issue after 7 days.

Cryoris commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+        layer_idx += 1
+
+        # check if every edge has been used at-least `min_2q_gate_per_edge` number of times.
+        reached_depth = np.array(list(edges_used.values())) >= min_2q_gate_per_edge
```
What happens if an edge has weight 0?

MozammilQ replied:
It raises a `ValueError` L230 of this file.

Cryoris commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,394 @@
+    insert_1q_oper=False,
+    prob_conditional: float = 0.1,
+):
+    """Generate random circuit of arbitrary size and form which strictly respects
```
It also seems like there could be a significant variance in circuit depth for graphs that have very varying edge weights. How about using the total number of gates as a standard termination criterion? Otherwise it would be important to clearly point out what the termination criterion is 🙂

MozammilQ replied:
@Cryoris , 
Couldn't agree more :) 

In that spirit, I have added an argument `rel_diff` read the docstring for more information.

MozammilQ replied:
![rel_diff](https://github.com/Qiskit/qiskit/assets/54737489/f3563f68-b4b9-4c1f-93a7-36939e601517)

MozammilQ replied:
Although its not a linear knob, but at-least this can be used to keep check on the size of the circuit generated.

I know these generated circuits are supposed to be fed to transpiler running on Rust,  which is really fast.
But, in case if a user wants to draw the circuits, this would be a lot of overhead, if the number of gates, especially measure is more.

Let me know what the reviewer thinks, I have added the code for `rel_diff` , added the docstring, and, mentioned in the release note. I will remove those on the sayings of the reviewer.

MozammilQ replied:
How `rel_diff` works,

Post normalization of the edge weights, I have taken a relative difference between the minimum and maximum probabilities, because this is the variable that would determine the depth of the circuit.

A user can keep a check on the circuit size, by controlling this variable `rel_diff`.

Cryoris replied:
I think it's better to keep the function as simple as possible, especially on last minute changes. It might be best to remove `rel_diff` and just clearly state how this function chooses the termination 🙂

1ucian0 commented:
> I am committing some unexpectedly silly mistakes here, will come back to this issue after 7 days.

FYI, The PR needs to be merged in the two coming days to be eligible for the Unitary Hack bounty. We are happy to continue working on it independently of it, but I just want to make sure you are aware.

MozammilQ commented:
> FYI, The PR needs to be merged in the two coming days to be eligible for the Unitary Hack bounty. We are happy to continue working on it independently of it, but I just want to make sure you are aware.

Although, a lame excuse, but its true I was not aware of this :)

MozammilQ committed:
added suggestions

MozammilQ committed:
refactor tests; added tests; lint

MozammilQ committed:
added a few more tests

MozammilQ committed:
added rel_diff; altered docstrings; added tests

MozammilQ commented:
For the last 4 commit, All the tests have been passed at my local system.

MozammilQ commented:
@sbrandhsn , please see if this is good enough :)

MozammilQ committed:
Merge branch 'Qiskit:main' into add_random_circuit_with_graph

MozammilQ committed:
restoring symengine version

MozammilQ committed:
added rel_diff to docstrings; releasenote; lint

MozammilQ commented:
Waiting for sabastian's view

On Wed, 19 Jun, 2024, 6:12 pm Julien Gacon, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In qiskit/circuit/random/utils.py
> <https://github.com/Qiskit/qiskit/pull/12474#discussion_r1646134511>:
>
> > +    ]
> +    return (gates_1q, gates_2q, gates_3q, gates_4q)
> +
> +
> +def random_circuit_from_graph(
> +    interaction_graph,
> +    min_2q_gate_per_edge,
> +    max_operands=2,
> +    measure=False,
> +    conditional=False,
> +    reset=False,
> +    seed=None,
> +    insert_1q_oper=False,
> +    prob_conditional: float = 0.1,
> +):
> +    """Generate random circuit of arbitrary size and form which strictly respects
>
> I think it's better to keep the function as simple as possible, especially
> on last minute changes. It might be best to remove rel_diff and just
> clearly state how this function chooses the termination 🙂
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Qiskit/qiskit/pull/12474#discussion_r1646134511>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANBTUUJ7NQLFZV3TLBQ7WJDZIF4FHAVCNFSM6AAAAABIN5ENOGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCMRYGE3TOMJVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

sbrandhsn commented:
Thanks for raising the point on circuit size @Cryoris @MozammilQ and sorry for the long delay on this code review @MozammilQ !
There seems to be two user expectations on what this function should provide:
* A user may expect the output quantum circuit to contain two-qubit gates that only have been drawn from the input interaction graph without the guarantee that the interaction graph of the output quantum circuit actually matches the input interaction graph. Here the regular definition of depth as given in the default `random_circuit` function may be used to limit the depth of the output circuit.
* A user may expect the interaction graph of the output quantum circuit to exactly match the input interaction graph. In this picture, a user may want a minimum number of two-qubit gates per edge in the input interaction graph. This may lead to an extremely deep quantum circuit. 

I think users of the default `random_circuit` method would probably expect the first variant whereas people working on e.g. circuit cutting or distributed quantum circuits might want/expect the second variant. The way to resolve this is to have two parameters that are mutually exclusive: `min_gate_per_edge` and 'depth', where the behavior of the first parameter is as set in the specification of the issue and the second parameter works as given in `random_circuit`. I think it is up to @MozammilQ; do you want continue on this PR that implements above behavior or do you want to implement the original specification that includes something like `min_gate_per_edge` and leave a `depth` parameter to future work? :-)

I like the garbage-in-garbage-out principle, so I'm not too keen on having a `rel_diff` parameter - if a user decides to set `min_gate_per_edge` and a skewed edge probability distribution, it is going to be on them when the resulting circuit is very deep... :-)

MozammilQ commented:
I feel a bit of guilt with this PR, because the reviewer @sbrandhsn  is being poked again and again for a review, for all my last PRs review has taken time, and Scientists are really busy that's a fact!
So, no need to be sorry @sbrandhsn , all reviewers have given me ample time and reviewers deserve even more time to make sure quality code gets merged :) 

I will implement the original specification that includes `min_gate_per_edge`.
and, I would be more than happy to work on `depth` and any/many other enhancements in future PRs.

and, I will remove `rel_diff` its related codes blocks.

sbrandhsn commented:
No worries, sounds good! :-)

MozammilQ commented:
I feel like a owe an explanation here;

1>> This PR's target user are the ones who are working on circuit cutting or distributed quantum circuits, the edges onto which 2Q gates are applied will be the edges present in the input interaction graph.

I believe that the current status/behavior of the PR is in line with the `1>>`.

So, the only change required now, is the removal of `rel_diff` and the PR should be ready for review :)

MozammilQ committed:
removed rel_diff

MozammilQ commented:
@sbrandhsn , I believe the PR is already inline with your interests.

I have just removed `rel_diff`.

Please, have a look and tell me what to change now :) 
It looks like, I am already not eligible for any bounty, so take your time :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        (standard_gates.C3SXGate, 4, 0),
+        (standard_gates.RC3XGate, 4, 0),
+    ]
+    return (gates_1q, gates_2q, gates_3q, gates_4q)
```
This can be simplified significantly by iterating over https://github.com/Qiskit/qiskit/blob/63b459eb125949fb7602a8617b56c71e1500f0be/qiskit/circuit/library/standard_gates/__init__.py#L49

It would also make the code easier to use and would include new standard gates by automatically.

sbrandhsn replied:
Also, you could make `_get_gates` accept an integer parameter `n` to select the gates with `n` qubits.

MozammilQ replied:
Done :)

MozammilQ replied:
I have chosen not to return gates as `np.ndarray`, because it could happen that `_get_gates()` is used in some future functions where `np.ndarray` is not required?

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    are chosen at random, the qubit-pairs are also chosen at random based on the probability
+    attached to the qubit-pair, the 2Q gates are applied on the qubit-paris on which are no gates
+    already present for that particular iteration, this is to make sure that for a particular
+    iteration only one circuit layer exists, now for those qubits for this particular iteration on
```
I don't understand this section. If you add gates to qubit-pairs that are not present in the particular iteration, you would end up with a completely dense layer where some qubit-pairs might not be part of the input interaction graph, right?

MozammilQ replied:
1>> extract qubit-pairs from input interaction graph

2>> Iterate over these qubit-pairs, apply 2Q gates but make sure those 2Q gates are
only applied to those qubit-pairs which are idle(both control and target), in case
any control, target found not to be idle, skip to another qubit-pair.

3>> Now, if any qubit still remains idle, apply 1Q if `insert_1Q_oper` is enabled.

A qubit which has only reset applied to it is considered idle.
Anyways, provided a better explanation in the docstring.

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    already present for that particular iteration, this is to make sure that for a particular
+    iteration only one circuit layer exists, now for those qubits for this particular iteration on
+    which there are no 2Q gates already applied, if `insert_1q_oper` is set to True, randomly
+    chosen 1Q gates are applied.
```
```suggestion
    which there are no 2Q gates already applied, if `insert_1q_oper` is set to True, randomly
    chosen 1Q gates are applied to qubits that are idle in a layer.
```

MozammilQ replied:
Done :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    relative difference means there is huge difference in the probabilistic selection of the edges,
+    this would force more iterations to happen just to find that edge whose probability of getting
+    selected is comparatively low, thereby increasing the depth of the circuit, and thus the whole
+    size of the circuit.
```
This can be removed, right?

MozammilQ replied:
Yup :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+       qc.draw(output='mpl')
+
+    Args:
+        interaction_graph (int): Interaction Graph
```
Why is `interaction_graph` of type `int`? `interaction_graph` should be a graph, preferably an undirected one

MozammilQ replied:
All the time I was thinking the input `interaction_graph` as the output of `build_interaction_graph` https://github.com/Qiskit/qiskit/blob/6a872a8e48e642631c73ee37ec4b949d98386d16/qiskit/transpiler/passes/layout/vf2_utils.py#L29

Now, I know its a graph, and not a tuple.
Sorry!
Graph could be PyDiGraph or PyGraph ( as you said undirected )

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        reset (bool): if True, insert middle resets. ( insert_1q_oper should be true )
+        (optional, default: False)
+        seed (int): sets random seed. (If `None`, a random seed is chosen) (optional)
+        insert_1q_oper (bool): Insert 1Q operations to the circuit. (optional, default: False)
```
I think a default of `True` makes sense for `insert_1q_oper`.

MozammilQ replied:
Done :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        CircuitError: When `max_operands` is not 1 or 2.
+        CircuitError: When `reset` is set to true, but no 1Q operations are allowed by setting
+        `insert_1q_oper` to false.
+        ValueError: when any edge have probability as zero or some are None but not all.
```
I think a probability of 0 should be fine, just remove the edge from the input interaction graph...

MozammilQ replied:
Nice! :) 
How could I not think about this?? 
Done :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    """
+
+    # If conditionals are set but not 1Q operation are allowed, raise an error in that case.
+    if reset and not insert_1q_oper:
```
I think we can allow setting `reset` and not setting `insert_1q_oper` the semantics would be that we only insert reset operations.

MozammilQ replied:
Done :) 
Now, `reset` is independent of `insert_1q_oper`

An idle qubit which has only reset applied on it is considered idle.

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        or, any of the probabilities are negatve.
+    """
+
+    # If conditionals are set but not 1Q operation are allowed, raise an error in that case.
```
```suggestion
    # If reset are set but not 1Q operation are allowed, raise an error in that case.
```

MozammilQ replied:
Foolish, typo :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        )
+
+    # Leaving out 'free_nodes'
+    pydi_graph, pydi_graph_node_map, _, _ = interaction_graph
```
`interaction_graph` should be an undirected graph, not a tuple. Node 0 in the input interaction graph should correspond to qubit 0 in the circuit.

MozammilQ replied:
All the time I was thinking the input `interaction_graph` as the output of `build_interaction_graph` https://github.com/Qiskit/qiskit/blob/6a872a8e48e642631c73ee37ec4b949d98386d16/qiskit/transpiler/passes/layout/vf2_utils.py#L29

Now, I know its a graph, and not a tuple.
Sorry!
Graph could be PyDiGraph or PyGraph ( as you said undirected )

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    if prob_weighted_mapping:
+        for prob in edges_probs:
+            if prob < 0:
+                raise ValueError("The probability should be positive")
```
You can move this check into the body of the first `if` above to simplify your code, then you can remove `prob_weighted_mapping` altogether.

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        # The number of vertices in interaction_graph define the number of qubits
+        # in the circuit to be generated
+        qubits = list(pydi_graph_node_map.keys())
+    else:
```
`interaction_graph` should only be an undirected graph, not a tuple. Changing this would remove this code.

MozammilQ replied:
Done :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+    else:
+        n_q = pydi_graph.num_nodes()
+        qubits = QuantumRegister(n_q)._bits
+    num_qubits = len(qubits)
```
I would strongly suggest to just create a quantum circuit  with the correct number of qubits here

MozammilQ replied:
This code block was a result of bag logic,
Got rid of this :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+            edge_list,
+            size=num_edges,
+            replace=True,
+            p=edges_probs if prob_weighted_mapping else None,
```
At this point, `edge_probs` should always be set to some value.

MozammilQ replied:
Yes

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        cumsum_params = np.cumsum(gate_choices["num_params"], dtype=np.int64)
+        parameters = rng.uniform(0, 2 * np.pi, size=cumsum_params[-1])
+
+        if conditional and layer_idx != 0:
```
Why would you disallow conditional gates in the 0th layer? Because they would require a previous measurement? I think you do not need the check here. In the broader picture of classical variable capabilities, a user may set the value of classical bits arbitrarily in the beginning of the circuit...

MozammilQ replied:
Removed skipping 1st layer :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+                    )
+                    qubit_idx_used.update(set(edge))
+                    qubit_idx_not_used = qubit_idx_not_used - qubit_idx_used
+                    edges_used[(control_qubit, target_qubit)] += 1
```
I think this branch is not necessary, you can use the logic for `conditionals` and `layer_idx>0`.

MozammilQ replied:
This would led to the calculation of the boolean list of conditionals, even if conditionals is not required :)
This would add a tiny bit of overhead which could have been avoided if this block doesn't run at all for conditional set to False.
That is why I had kept the two separate blocks :)

Anyways, change the code to your liking :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+            condition_values = rng.integers(
+                0, 1 << min(num_qubits, 63), size=np.count_nonzero(is_conditional)
+            )
+            c_ptr = 0
```
`c_ptr` sounds like a weird variable name, I would suggest to use something like `clbit_idx`

MozammilQ replied:
Agree :)

sbrandhsn commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,415 @@
+        layer_idx += 1
+
+        # check if every edge has been used at-least `min_2q_gate_per_edge` number of times.
+        reached_depth = np.array(list(edges_used.values())) >= min_2q_gate_per_edge
```
Could this line be the `while` condition in line 285? If yes, you can remove variable `stop` and make the condition clearer with a comment in line 284.

MozammilQ replied:
I wanted to achieve a do-while loop behavior,
There is no need to check the condition for the first time, because its certain that all the qubits would be idle.
That's why I had checked the condition at the very last, and to make the code look pretty used a variable `stop`.

Anyways, changed the code to your liking :)

sbrandhsn commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,60 @@
+    the edge of the graph generated. If all the probabilities is passed as `None`, then the probability
+    of each qubit-pair of getting selected is set to 1/N. (where N is the number of edges in the 
+    interaction_graph passed in) 
+    If float values are present as probabilities but some are None, this will raise a ValueError.
```
```suggestion
    The probability of randomly drawing an edge from the interaction graph as a two-qubit gate can be set by the user in the weight attribute of an edge in the input interaction graph. If the user does not set the probability, each edge is drawn uniformly, i.e. each two-qubit gate represented by an edge in the interaction graph has the same probability of getting added to the random circuit. If only a subset of edge probabilities are set, `ValueError` will be set.
```

MozammilQ replied:
Done :)

sbrandhsn commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,60 @@
+features_circuits:
+  - |
+    The function :func:`~qiskit.circuit.random.utils.random_circuit_from_graph` adds a new feature of 
+    generating random circuits with interaction graph.
```
```suggestion
    The function :func:`~qiskit.circuit.random.utils.random_circuit_from_graph` generates a random circuit that induces the same interaction graph as the interaction graph specified by `interaction_graph`.
```

MozammilQ replied:
Done :)

sbrandhsn reviewed (dismissed):
Thanks @MozammilQ ! I left a couple of comments - it would be great if you can address them. I like that you improved the code with a lot of comments compared to my last review!

MozammilQ committed:
added suggestion; refactor code

MozammilQ committed:
added a bit of explanation; removed delay

MozammilQ commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -160,3 +165,241 @@ def test_random_circuit_with_zero_distribution(self):
+        for cp in cp_mp:
+            self.assertTrue(cp in edge_list)
+
+    def test_2q_gates_excluded_edges_with_zero_weight(self):
```
I am doing code duplication because I want to take a test `cp_map` with zero weight for this test only.

MozammilQ commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,333 @@
+
+    # If any edge weight is zero, just remove that edge from the edge_list
+    if 0 in edges_probs:
+        edge_list = [edge for edge, edge_prob in zip(edge_list, edges_probs) if not edge_prob == 0]
```
Choosing this logic over "pydi_graph.remove_edge"

MozammilQ commented:
@sbrandhsn , please see if its good enough :)

MozammilQ committed:
altered comments

MozammilQ commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,335 @@
 
 
+def _get_gates(n_qubits):
+    return [
```
I have chosen not to return gates as np.ndarray, because it could happen that `_get_gates()` is used in some future functions where ndarray is not required?

MozammilQ replied:
This is not efficient, it calls the same function as many times as the number of qubits in the gates are allowed.

I still have kept the changes because my opinion is that all the standard gates and closely related stuffs should be in one place, and any changes to them should reflect in all corners of Qiskit.

I am thinking of doing some changes to `qiskit.circuit.library.standard_gates.__init__.py` but that should be out of scope of this PR.

Now, regarding that 1 failing test, If I do not use `get_standard_gate_name_mapping` and do it in the old way, it passes.
But, If I use `get_standard_gate_name_mapping` and change instr.param, or even if do `instr.__init__(*args)` the test fails, it used to pass couple of months ago, but I don't know why when qpy loads the circuit the parameters changes somehow, and that fails the test.

@eliarbel , expecting your valuable comments :)

eliarbel replied:
Looking a bit at the failure, I'm afraid that your instruction-based implementation doesn't play well here with the singleton and caching models implemented for circuit data in the Rust space. In particular, the parameter updates of mutable instructions like here: https://github.com/Qiskit/qiskit/blob/7cb79aaac5bfe501367ca07f94ab03ba64c51b68/qiskit/circuit/random/utils.py#L620 end up with the instructions cached in Rust inconsistent with what you see on the Python side. Furthermore, QPY traverses the `QuantumCircuit`'s Rust data structure, hence the difference. 

I think the best path forward is to resort to the implementation paradigm we had before in `random_circuit` , i.e. with the `Gate` classes lists. Practically, I would revert the changed in `random_circuit` and change the implementation in `random_circuit_from_graph` to follow a similar approach. In general BTW I liked your idea of pulling the standard gate lists from one place (as with `get_standard_gate_name_mapping`), but this is something we can address in a separate PR.

MozammilQ replied:
@eliarbel , done with the changes, Please, see if the PR is good enough now :)

MozammilQ commented:
@sbrandhsn , I humbly request you to have a look at this one, only if you have time to do so :)

MozammilQ committed:
Merge branch 'Qiskit:main' into add_random_circuit_with_graph

MozammilQ commented:
@sbrandhsn ,  I humbly request you to have a look at this one, only if you have time to do so :)

MozammilQ committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ commented:
@sbrandhsn , @1ucian0 , things are being converted to rust these days, is that the reason the PR is on hold? If, yes should I go ahead and port it to rust?
If, not please have a look at this PR :) 

I will correct these errors if this PR is required anymore :)

MozammilQ committed:
Merge branch 'Qiskit:main' into add_random_circuit_with_graph

MozammilQ committed:
refactor code

MozammilQ committed:
refactor code

MozammilQ committed:
fix lint

MozammilQ committed:
optimized code; added a few tests

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    prob_conditional: float = 0.1,
+    prob_reset: float = 0.1,
+):
+    """Generate random circuit of arbitrary size and form which strictly respect the interaction
```
```suggestion
    """Generate random circuit of arbitrary size and form which strictly respects the interaction
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    If numerical values are present as probabilities but some/any are None, or negative, this will
+    raise a ValueError.
+
+    If :arg:`max_operands` is set to 1, then there are no 2Q operations, so no need to take care
```
```suggestion
    If `max_operands` is set to 1, then there are no 2Q operations, so no need to take care
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+
+    If :arg:`max_operands` is set to 1, then there are no 2Q operations, so no need to take care
+    of the edges, in such case the function will return a circuit from the `random_circuit` function,
+    which would be passed with the :arg:`max_operands` as 1.
```
```suggestion
    which would be passed with the `max_operands` as 1.
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    of the edges, in such case the function will return a circuit from the `random_circuit` function,
+    which would be passed with the :arg:`max_operands` as 1.
+
+    If :arg:`max_operands` is set to 2, then in every iteration `N` 2Q gates and qubit-pairs which
```
```suggestion
    If `max_operands` is set to 2, then in every iteration `N` 2Q gates and qubit-pairs which
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+        reset (bool): if True, insert middle resets. (optional, default: False)
+        seed (int): sets random seed. (If `None`, a random seed is chosen) (optional)
+        insert_1q_oper (bool): Insert 1Q operations to the circuit. (optional, default: True)
+        prob_conditional (float): Probability less than 1.0, this is used to control the occurence
```
```suggestion
        prob_conditional (float): Probability less than 1.0, this is used to control the occurrence
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+        insert_1q_oper (bool): Insert 1Q operations to the circuit. (optional, default: True)
+        prob_conditional (float): Probability less than 1.0, this is used to control the occurence
+        of conditionals in the circuit. (optional, default: 0.1)
+        prob_reset (float): Probability less than 1.0, this is used to control the occurence of
```
```suggestion
        prob_reset (float): Probability less than 1.0, this is used to control the occurrence of
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+        prob_conditional (float): Probability less than 1.0, this is used to control the occurence
+        of conditionals in the circuit. (optional, default: 0.1)
+        prob_reset (float): Probability less than 1.0, this is used to control the occurence of
+        reset in the cirucit. (optional, default: 0.1)
```
```suggestion
        reset in the circuit. (optional, default: 0.1)
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+        CircuitError: When the interaction graph has no edges, so only 1Q gates are possible in
+        the circuit, but `insert_1q_oper` is set to False.
+        ValueError: when any edge have probability None but not all or, any of the probabilities
+        are negatve.
```
```suggestion
        are negative.
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    # max_operands should be 1 or 2
+    if max_operands not in {1, 2}:
+        raise CircuitError(
+            "This function is intended to only work on single-qubit and two-qubit gates"
```
```suggestion
            "`max_operands` should be either 1 or 2"
```
I think this is a clearer message, mentioning the argument explicitly

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+
+    if num_edges == 0 and not insert_1q_oper:
+        raise CircuitError(
+            "There are no edges in the `interaction_graph` so, there could be no 2Q gates, "
```
```suggestion
            "There are no edges in the `interaction_graph` so there could be only 1Q gates, "
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    if num_edges == 0 and not insert_1q_oper:
+        raise CircuitError(
+            "There are no edges in the `interaction_graph` so, there could be no 2Q gates, "
+            "and only 1Q gates in the circuit, however `insert_1q_oper` is set to `False`."
```
```suggestion
            "however `insert_1q_oper` is set to `False`"
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    if all(edges_probs):
+        if not np.isclose(np.sum(edges_probs), 1.000, rtol=0.001):
+            # Normalize edge weights if not already normalized.
+            edges_probs = edges_probs / np.sum(edges_probs)
```
If all probabilities are negative, the negative sum will flip their sign back since it's negative. This does not match the behavior described in the docstring w.r.t negative probs.

MozammilQ replied:
Nice catch!
Have fixed the issue :)

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    # edge weights in interaction_graph must be positive
+    for prob in edges_probs:
+        if prob < 0:
+            raise ValueError("The probability should be positive")
```
```suggestion
            raise ValueError("Probabilities cannot be negative")
```

They can be zero

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+
+    edges_used = {edge: 0 for edge in edge_list}
+
+    # Declearing variables, so that lint doesn't complaint.
```
```suggestion
    # Declaring variables, so that lint doesn't complaint.
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+    cond_val_2q = None
+    cond_val_1q = None
+
+    # If conditional is not required, there is not need to calculate random numbers.
```
```suggestion
    # If conditional is not required, there is no need to calculate random numbers.
```

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+        cumsum_params = np.cumsum(gate_choices["num_params"], dtype=np.int64)
+        parameters = rng.uniform(0, 2 * np.pi, size=cumsum_params[-1])
+
+        # If reset is required, then, Generating a random boolean matrix of
```
```suggestion
        # If reset is required, then, generating a random boolean matrix of
```

eliarbel commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,71 @@
+---
+features_circuits:
+  - |
+    The function :func:`~qiskit.circuit.random.utils.random_circuit_from_graph` generates a random circuit that 
```
```suggestion
    Added a function :func:`~qiskit.circuit.random.utils.random_circuit_from_graph` that generates a random circuit that 
```
More in line with how we document new functionality in the release notes

Also, please update `qiskit/circuit/__init__.py` to include the function in the api docs, and make sure the link here works when you render locally the release notes (with `tox -e docs`)

MozammilQ replied:
The changes are in [ebe1de0](https://github.com/Qiskit/qiskit/pull/12474/commits/ebe1de0072977e4543dac8927e2802f599818201)
I have followed exactly what was mentioned in "Generating random circuits" section of `qiskit/circuit/__init__.py`

eliarbel commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,71 @@
+    same probability of getting added to the random circuit. If only a subset of edge probabilities are set, 
+    `ValueError` will be raised.
+
+    Arguments accepted by the function are
```
I would remove the arguments paragraph, seems too much for a release note (users can go to the docstring itself for more details).

MozammilQ replied:
Right, done !

eliarbel commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,71 @@
+    :code:`prob_reset` A float value less than 1.0, this is used to control the occurence of resets. 
+    (default 0.1) (optional)
+
+    This is an example where 'cp_map' is a list of edges for a "Directed Heavy Hex Graph" with some arbitrary weights.
```
I would put a simpler example here (say the one in the function's docstring or even simpler). 
Also, please use `.. plot::` instead of `code-block`

MozammilQ replied:
Done!

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+
+
+def random_circuit_from_graph(
+    interaction_graph,
```
I think it would be very nice from user experience to also allow a user to just send a list of tuples, say `[(0, 1, 1), (1, 2, 1), (2, 3, 2)]` and have the function handle itself the rustworkx construction (inferring the number of nodes automatically in this case, based on the edges and say filling up holes in node indices, if there are any).

MozammilQ replied:
Just to adhere with the instruction in the issue, I decided to keep it as `PyDiGraph`
> "The [generation of random circuits](https://github.com/Qiskit/qiskit/blob/main/qiskit/circuit/random/utils.py) should be extended by the ability to specify a circuit connectivity or interaction graph G that the random graph should should adhere to."

The requested changes are in [eab2a4c](https://github.com/Qiskit/qiskit/pull/12474/commits/eab2a4c191839518ad554ab8bfcad23370405084)

eliarbel replied:
Thanks for adding this, I already used it in my review testing 💯

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,16 +11,385 @@
+
+    # This loop will keep on applying gates to qubits until every qubit-pair
+    # has 2Q operations applied at-least `min_2q_gate_per_edge` times.
+    while not all(np.array(list(edges_used.values())) >= min_2q_gate_per_edge):
```
I'm wondering, do we want to exclude an edge from being assigned a random 2q gate once the `min_2q_gate_per_edge` constraint is satisfied for it? Maybe as an option? Not sure what is the required behavior here, so this can be ignored.

eliarbel reviewed (changes requested):
Thanks @MozammilQ for your patience regarding a follow-up review. 
Overall this seems in line with the spec in https://github.com/Qiskit/qiskit/issues/12364. There are some changes required before it can be merged, please see my line comments. 

Also:
1. Regarding testing: I'm missing some validation related to the probabilities aspect. Can you come up with a test(s) (possibly using a fixed seed) to make sure 2Q gate distribution is within a reasonable expectancy w.r.t to the input probabilities?  
2. Performance: for very large circuits, how is the `random_circuit_from_graph` performance compared to `random_circuit`?

MozammilQ committed:
fixed many typos; added suggestions by Eli Arbel

MozammilQ committed:
added random circuit API docs

MozammilQ committed:
fixed releasenote; added test to validate probs

MozammilQ committed:
added support of cp_map; fixed lint

MozammilQ commented:
@eliarbel , I have done with the changes, please do have a look :) 

 > 
 > 1. Regarding testing: I'm missing some validation related to the probabilities aspect. 
 >      Can you come up with a test(s) (possibly using a fixed seed) to make sure 2Q 
>       gate distribution is within a reasonable expectancy w.r.t to the input probabilities?
> 2. Performance: for very large circuits, how is the random_circuit_from_graph 
        performance compared to random_circuit?
 

The test has been added in [6e7ede8](https://github.com/Qiskit/qiskit/pull/12474/commits/6e7ede8c7696aea89a4d0b3e6929f90cae98a6e7)

Below is the performance comparison :) 
I am comparing the performance based on time taken to generate same number of operations.


```python
from qiskit.circuit.random import random_circuit_from_graph, random_circuit
import numpy as np
import rustworkx as rx
import time

distance = 5
h_h_g = rx.generators.directed_heavy_hex_graph(d=distance, bidirectional = False)
seed = 32434
rng = np.random.default_rng(seed = seed)
cp_map_list = []
edge_list = h_h_g.edge_list()

# generating a non-normalized list of integers.
random_probs = rng.choice(range(10, 25), size = len(edge_list)).tolist()
sum_probs = sum(random_probs)

for idx, qubits in enumerate(edge_list):
    ctrl, trgt = qubits
    cp_map_list.append((ctrl, trgt, random_probs[idx]))

h_h_g.clear_edges()
h_h_g.add_edges_from(cp_map_list)

start_time = time.time()
qc = random_circuit_from_graph(h_h_g,
                               min_2q_gate_per_edge = 10,
                               max_operands = 2,
                               measure = True,
                               conditional = True, # Just making it a bit more challenging.
                               reset = True,
                               seed = seed,
                               insert_1q_oper = True,
                               prob_conditional = 0.91,
                               prob_reset = 0.50)
end_time = time.time()
count_oper = sum(qc.count_ops().values())
print(f"For `random_circuit_from_graph`:\nTime taken: {end_time - start_time} to generate {count_oper} operations")
print(f"Time taken per operation: {(end_time-start_time)/count_oper*1000000} us\n", 40*'#', '\n')

rand_start_time = time.time()
qc_rand = random_circuit(num_qubits = 70, depth = 299, max_operands = 2, measure = True, conditional = True, reset = True, seed = seed)
rand_end_time = time.time()
count_rand_oper = sum(qc_rand.count_ops().values())
print(f"For `random_circuit`:\nTime taken: {rand_end_time-rand_start_time} to generate {count_rand_oper} operations")
print(f"Time taken per operation: {(rand_end_time-rand_start_time)/count_rand_oper*1000000} us")
```
**Output**
```text
For `random_circuit_from_graph`:
Time taken: 18.229220867156982 to generate 151169 operations
Time taken per operation: 120.58835387650234 us
 ######################################## 

For `random_circuit`:
Time taken: 18.196982622146606 to generate 151748 operations
Time taken per operation: 119.91579870671512 us
```

MozammilQ committed:
Merge branch 'Qiskit:main' into add_random_circuit_with_graph

MozammilQ committed:
refactor code

MozammilQ committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ committed:
removed c_if dependency

eliarbel commented on qiskit/circuit/__init__.py:
```diff
@@ -306,6 +306,8 @@
 There is also a utility for generating random circuits:
```
```suggestion
There are also utilities for generating random circuits:
```

MozammilQ replied:
Done

eliarbel commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,35 @@
+    This is an example where 'cp_map' is a list of edges with some arbitrary weights.
+
+    .. plot::
+       
```
```suggestion
             :include-source:
```
This will render the code also

MozammilQ replied:
Done

eliarbel commented on releasenotes/notes/Added-random-circuit-from-graph-95c22eeabdea89d0.yaml:
```diff
@@ -0,0 +1,35 @@
+       pydi_graph.add_edges_from(cp_map)
+       # cp_map can be passed in directly as interaction_graph
+       qc = random_circuit_from_graph(interaction_graph = pydi_graph, 
+       min_2q_gate_per_edge = 1, 
```
I would indent the arguments lines for better readability

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/__init__.py:
```diff
@@ -10,6 +10,46 @@
+Overview
+========
+
+Existing architecture of Quantum Computers have varying computational capabilities.
```
I think it's a good move introducing the `qiskit.circuit.random` module now 👍 

I would go for a bit more module-specific overview text here. Something like:

The :mod:`qiskit.circuit.random` module offers functions that can be used for generating arbitrary circuits with
gates randomly selected from a given set of gates. These circuits can be used for
benchmarking existing quantum hardware and estimating the performance of quantum
circuit transpilers and software infrastructure.
The functions below can generate bespoke quantum circuits respecting various properties
such as number of qubits, depth of the circuit, coupling map, gate set, etc.

MozammilQ replied:
Done !

eliarbel commented on qiskit/circuit/random/__init__.py:
```diff
@@ -10,6 +10,46 @@
+.. autofunction:: random_circuit_from_graph
+
+
+Generating arbitrary circuits
```
Please make this appear before the `random_circuit_from_graph` entry. I think it's better to start with the most generic function in the module

MozammilQ replied:
Done !

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+
+
+def random_circuit_from_graph(
+    interaction_graph,
```
Do we want type hints here?

MozammilQ replied:
Actually, no!
Type hints are already present in the docstring.

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+
+def random_circuit_from_graph(
+    interaction_graph,
+    min_2q_gate_per_edge: int,
```
I would have a default value of 1 here

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    circuit, and, E is the set of two-qubit gate interactions between two particular qubits in the
+    circuit.
+
+    This function will generate a random circuit by randomly selecting gates from the set of
```
```suggestion
    This function will generate a random circuit by randomly selecting 1Q and 2Q gates from the set of
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    circuit.
+
+    This function will generate a random circuit by randomly selecting gates from the set of
+    standard gates in :mod:`qiskit.circuit.library.standard_gates`. User can attach a numerical
```
```suggestion
    standard gates in :mod:`qiskit.circuit.library.standard_gates`. The user can attach a numerical
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    This function will generate a random circuit by randomly selecting gates from the set of
+    standard gates in :mod:`qiskit.circuit.library.standard_gates`. User can attach a numerical
+    value as a metadata to the edge of the graph indicating the edge weight for that particular edge,
+    These edge weights would be normalized to the probabilities for the edges of getting selected.
```
```suggestion
    These edge weights would be normalized to the probabilities of the edges getting selected.
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    value as a metadata to the edge of the graph indicating the edge weight for that particular edge,
+    These edge weights would be normalized to the probabilities for the edges of getting selected.
+    If all the edge weights are passed as `None`, then the probability of each qubit-pair of getting
+    selected is set to 1/N, where `N` is the number of edges in the interaction_graph passed in,
```
```suggestion
    selected is set to 1/N, where `N` is the number of edges in the `interaction_graph` passed in,
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    If all the edge weights are passed as `None`, then the probability of each qubit-pair of getting
+    selected is set to 1/N, where `N` is the number of edges in the interaction_graph passed in,
+    i.e each edge is drawn uniformly. If any weight of an edge is set as zero, that particular
+    edge will be not be included in the output circuit.
```
```suggestion
    edge will not be included in the output circuit.
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+    edge will be not be included in the output circuit.
+
+    Passing a list of tuples of control qubit, target qubit and associated probability is also
+    acceptable.
```
```suggestion
    acceptable as a way to specify an interaction graph.
```

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+                num_qubits = trgt
+
+        num_qubits += 1  # ctrl, trgt are qubit indices.
+    else:
```
I would make this something like `elif isinstance(interaction_graph, rustworkx.PyDiGraph):` and raise `CircuitError` in the `else` branch to indicate an invalid interaction graph object

MozammilQ replied:
Done

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+
+        # normalized edge weights represent the probability with which each qubit-pair
+        # interaction is inserted into a layer.
+        edge_choices = rng.choice(
```
I think there is another optimization opportunity here with `edge_choices`:
In line 339: 
```python
if control_qubit in qubit_idx_not_used and target_qubit in qubit_idx_not_used:
```
You skip edges whose qubits were already used for the current layer. So instead skipping used qubits like that you can remove already here the edges in `edge_choices` which will result with a skip. This will save preparation work in terms randomly choosing gates and parameters below. Then you can also get rid of maintaining `qubit_idx_used` and `qubit_idx_not_used` inside the `while` loop. 
In a quick check I did I found that its not uncommon to have only third of the instructions actually being added to the circuit while the rest are just skipped because of the if check above.

MozammilQ replied:
Removed `qubit_idx_used`. `qubit_idx_not_used` changed to `qubit_idx_idle` this name better represents its function.
`qubit_idx_idle` has to exist because it's used in the `if insert_1q_oper` block

> In a quick check I did I found that its not uncommon to have only third of the instructions actually being added to the circuit while the rest are just skipped because of the if check above.
That is so inefficient!

Anyways, I have realized that skipping any edge that has been probabilistically chosen, meddles with the accuracy of the probability of the edges that are being added to the circuit, as you can see in the `test.python.circuit.test_random_circuit.TestRandomCircuitFromGraph.test_edges_probs` that I have achieved 1% accuracy with more deviation in probability of the edges in the `interaction_graph` that too with generation of less computationally expensive circuit.

This time whatever edges that are chosen in `edge_choices` straightaway goes to the circuit, so the edges in the circuit maintain the probability exactly equal to the ones passed in the `interaction_graph`.
Now, to comply with `min_2q_gate_per_edge` and the termination of the loop. A record of the edges with its frequency of occurrences in the circuit has been maintained in `edges_used` and those that have attained `min_2q_gate_per_edge` is removed from the record `edges_used`, the while-loop terminates when all the edges are removed from `edge_used` marking the frequency of all edges in the circuit equal to `min_2q_oper_per_edge`.

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -11,15 +11,437 @@
+
+    # This loop will keep on applying gates to qubits until every qubit-pair
+    # has 2Q operations applied at-least `min_2q_gate_per_edge` times.
+    while not all(np.array(list(edges_used.values())) >= min_2q_gate_per_edge):
```
From optimization perspective, this operation seems unnecessary to me. Maybe instead try something like just checking:
```python
while edges_used:
``` 
And inside the `while` loop remove an edge from the `edges_used` dictionary as soon as it meets the `min_2q_gate_per_edge` requirement.

MozammilQ replied:
I have changed the logic, this time, the aim is to attain 'min_2q_gate_per_edge' as fast as possible per iteration, this leads to more 2Q gates on the circuit with less preparation.
this also means that 1 iteration of the while-loop doesn't correspond to 1 layer in the circuit.
inside the `for` loop  I am removing an edge from the `edges_used` dictionary as soon as it meets the `min_2q_gate_per_edge` requirement.

Also see [comment](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2024510973)

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -53,11 +475,10 @@ def random_circuit(
-            deviation from the specified ratios that depends on the size of the requested
-            random circuit. (optional)
-
+        num_operand_distribution (dict): a distribution of gates that specifies the ratio of 1-qubit,
```
We normally indent multiline arguments text in docstring just once. Please revert to the original convention. 

Same goes for the args of `random_circuit_from_graph`.

MozammilQ replied:
Done!

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -314,7 +690,7 @@ def random_clifford_circuit(num_qubits, num_gates, gates="all", seed=None):
         num_gates (int): number of gates in the circuit.
         gates (list[str]): optional list of Clifford gate names to randomly sample from.
-            If ``"all"`` (default), use all Clifford gates in the standard library.
+                           If ``"all"`` (default), use all Clifford gates in the standard library.
```
Please revert to the original indentation.

MozammilQ replied:
Done :)

eliarbel commented on qiskit/circuit/random/utils.py:
```diff
@@ -262,33 +635,36 @@ def random_circuit(
                 else:
-                    qc._append(
-                        CircuitInstruction(operation=operation, qubits=qubits[q_start:q_end])
+                    qc.append(
```
`qc._append` is faster since it skips validations which are done in `qc.append` so is preferred. In general I would revert the  code block inside `if conditional and layer_number != 0:` to its original state (unless I'm missing the purpose of the name changes there).

MozammilQ replied:
Reverted back the name change, regarding `qc.append` see [comment](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2024522147)

eliarbel commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -75,13 +80,14 @@ def test_random_mid_circuit_measure_conditional(self):
         circ = random_circuit(num_qubits, depth, conditional=True, seed=16)
         self.assertEqual(circ.width(), 2 * num_qubits)
         op_names = [instruction.operation.name for instruction in circ]
+        self.assertEqual(4, len(op_names))
```
What changed in `random_circuit` that requires this change?

MozammilQ replied:
Referring to `test.python.circuit.test_random_circuit.TestCircuitRandom.test_random_mid_circuit_measure_conditional`

`qc.append` leads to the following result.

For this code:
```python
from qiskit.circuit.random import random_circuit
num_qubits = depth = 2
circ = random_circuit(num_qubits, depth, conditional=True, seed=16)
for instr in circ._data:
    print(instr)
```

```text
CircuitInstruction(operation=Instruction(name='ry', num_qubits=1, num_clbits=0, params=[np.float64(3.9050564582327985)]), qubits=(<Qubit register=(2, "q"), index=0>,), clbits=())
CircuitInstruction(operation=Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), qubits=(<Qubit register=(2, "q"), index=0>,), clbits=(<Clbit register=(2, "c"), index=0>,))
CircuitInstruction(operation=Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), qubits=(<Qubit register=(2, "q"), index=1>,), clbits=(<Clbit register=(2, "c"), index=1>,))
CircuitInstruction(operation=Instruction(name='if_else', num_qubits=2, num_clbits=2, params=[<qiskit.circuit.quantumcircuit.QuantumCircuit object at 0x7d9dfe398710>, None]), qubits=(<Qubit register=(2, "q"), index=1>, <Qubit register=(2, "q"), index=0>), clbits=(<Clbit register=(2, "c"), index=0>, <Clbit register=(2, "c"), index=1>))
```
```python
circ.draw(output='mpl')
```
![with-if](https://github.com/user-attachments/assets/84996426-9b47-41ba-a047-5e45cc373392)


```python
len(op_names)
```
Gives: 4


Now, `qc._append` on the same code

For this code:
```python
from qiskit.circuit.random import random_circuit
num_qubits = depth = 2
circ = random_circuit(num_qubits, depth, conditional=True, seed=16)
for instr in circ._data:
    print(instr)
```

```text
CircuitInstruction(operation=Instruction(name='ry', num_qubits=1, num_clbits=0, params=[np.float64(3.9050564582327985)]), qubits=(<Qubit register=(2, "q"), index=0>,), clbits=())
CircuitInstruction(operation=Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), qubits=(<Qubit register=(2, "q"), index=0>,), clbits=(<Clbit register=(2, "c"), index=0>,))
CircuitInstruction(operation=Instruction(name='measure', num_qubits=1, num_clbits=1, params=[]), qubits=(<Qubit register=(2, "q"), index=1>,), clbits=(<Clbit register=(2, "c"), index=1>,))
CircuitInstruction(operation=Instruction(name='cs', num_qubits=2, num_clbits=0, params=[]), qubits=(<Qubit register=(2, "q"), index=1>, <Qubit register=(2, "q"), index=0>), clbits=())
CircuitInstruction(operation=Instruction(name='if_else', num_qubits=0, num_clbits=2, params=[<qiskit.circuit.quantumcircuit.QuantumCircuit object at 0x704da052f470>, None]), qubits=(), clbits=(<Clbit register=(2, "c"), index=0>, <Clbit register=(2, "c"), index=1>))
```

Noticed, the problem?

```python
for instr in circ._data:
    print(instr.operation.num_qubits)
```

Gives
```text
1
1
1
2
0
```

```python
circ.draw(output='mpl')
```
![no-if](https://github.com/user-attachments/assets/3e295077-adcb-452b-b2b8-4ba55d2fa38b)


```python
len(op_names)
```
Gives: 5


The above observation suggests that may be not in the future but at present we should go with `qc.append`.

Regarding the speed, `qc.append` looks ~x2 slower than `qc._append`.

eliarbel replied:
Ah... this was a good catch! So `QuantumCircuit._append` could not be used within a control-flow builder context (e.g. within `with qc.if_test`) since it doesn't look at the current circuit scope, effectively ignoring the control-flow scope here. This is why we don't see the if test surrounding the CS gate. 
So let's keep the `qc.append` inside all the control-flow build contexts here and use `qc._append` elsewhere. You can also do this in `random_circuit_with_graph` for an enhanced performance.

eliarbel commented on test/python/circuit/test_random_circuit.py:
```diff
@@ -160,3 +166,426 @@ def test_random_circuit_with_zero_distribution(self):
+        self.assertNotEqual(cond_counter_1q, 0)
+        self.assertNotEqual(cond_counter_2q, 0)
+
+    def test_edges_prob(self):
```
Thanks for adding this test. It turned out though to be a rather lengthy, about 33 seconds on my laptop. Can you try to make it faster, maybe by using 3 qubits in a way that still makes sense from the probabilities check perspective?

MozammilQ replied:
Done!

eliarbel replied:
6X faster now, thanks!

eliarbel reviewed (changes requested):
Thanks @MozammilQ for addressing the comments in the previous review, I think it was a big step forward. Now with `random_circuit` almost back to its original state, I gave a more thorough look at the other parts. Please see my line comments. I hope this will be the last round of changes before we can merge the PR.

Regarding your performance comparison, it was helpful. It does help with getting a rough sense for the scaling of `random_circuit_from_graph` compared to `random_circuit`. If we want at some point to get a better understanding of it, we'll need to explore the parameter space more rigorously. For example, I played with your code a bit and got this: 

```python
qc = random_circuit_from_graph(h_h_g,
                               min_2q_gate_per_edge = 90,
                               max_operands = 2,
                               measure = True,
                               conditional = False, # Just making it a bit more challenging.
                               reset = False,
                               seed = seed,
                               insert_1q_oper = True)

qc_rand = random_circuit(num_qubits = 70, depth = 298, max_operands = 2, measure = True, conditional = False, reset = False, seed = seed)

# For `random_circuit_from_graph`:
# Time taken: 1.6271507740020752 to generate 19729 operations
# Time taken per operation: 82.47507597962772 us
#  ########################################

# For `random_circuit`:
# Time taken: 0.5127830505371094 to generate 19067 operations
# Time taken per operation: 26.893745766880443 us
```
But still it's good to have some sense that the new function is not too far off. I've left a couple optimization suggestions that might help but I wouldn't go beyond that in this PR trying to optimize performance more. We can handle this aspect in a new PR if there will be a use-case for which performance becomes a bottleneck.

MozammilQ committed:
applied Eli Arbel's suggestion

MozammilQ committed:
fixed renormalization, typo

MozammilQ committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ commented:
Regarding performance code in the [comment](https://github.com/Qiskit/qiskit/pull/12474#pullrequestreview-2727333891)
Changed `distance` from 7 to 17
Changed `min_2q_gate_per_edge` from 90 to 200
Changed `num_qubits` from 70 to 600
and, `depth` from 298 to 620

After change in the logic [comment](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2024548261)
It seems like there is a x2 performance gain,

After building rust in release mode
```python
from qiskit.circuit.random import random_circuit_from_graph, random_circuit
import numpy as np
import rustworkx as rx
import time

distance = 17
h_h_g = rx.generators.directed_heavy_hex_graph(d=distance, bidirectional = False)
seed = 32434
rng = np.random.default_rng(seed = seed)
cp_map_list = []
edge_list = h_h_g.edge_list()

# generating a non-normalized list of integers.
random_probs = rng.choice(range(10, 25), size = len(edge_list)).tolist()
sum_probs = sum(random_probs)

for idx, qubits in enumerate(edge_list):
    ctrl, trgt = qubits
    cp_map_list.append((ctrl, trgt, random_probs[idx]))

h_h_g.clear_edges()
h_h_g.add_edges_from(cp_map_list)

start_time = time.time()
qc = random_circuit_from_graph(h_h_g,
                               min_2q_gate_per_edge = 200,
                               max_operands = 2,
                               measure = True,
                               conditional = False, 
                               reset = False,
                               seed = seed,
                               insert_1q_oper = True)

end_time = time.time()
count_oper = sum(qc.count_ops().values())
print(f"For `random_circuit_from_graph`:\nTime taken: {end_time - start_time} to generate {count_oper} operations")
print(f"Time taken per operation: {(end_time-start_time)/count_oper*1000000} us\n", 40*'#', '\n')

rand_start_time = time.time()
qc_rand = random_circuit(num_qubits = 600, depth = 620, max_operands = 2, measure = True, conditional = False, reset = False, seed = seed)
rand_end_time = time.time()
count_rand_oper = sum(qc_rand.count_ops().values())
print(f"For `random_circuit`:\nTime taken: {rand_end_time-rand_start_time} to generate {count_rand_oper} operations")
print(f"Time taken per operation: {(rand_end_time-rand_start_time)/count_rand_oper*1000000} us")
```

```text
For `random_circuit_from_graph`:
Time taken: 12.82305383682251 to generate 369606 operations
Time taken per operation: 34.69384651986848 us
 ######################################## 

For `random_circuit`:
Time taken: 6.758020639419556 to generate 339428 operations
Time taken per operation: 19.91002698486735 us
```

`random_circuit_from_graph` here with its parameters looks ~x1.6 slower than `random_circuit`, the observation remains consistent with the size of the circuit generated.

For logic change see [comment](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2024548261)
For `qc.append` see [comment](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2024522147)

Please, see if the changes are good enough :)
@eliarbel

MozammilQ committed:
refactor code; optimized test

eliarbel reviewed (changes requested):
Thanks @MozammilQ for making all the changes. Looks good! 
Regarding the `qc.append` comment, please see my response [here](https://github.com/Qiskit/qiskit/pull/12474#discussion_r2030892587). Let's apply it and we're done.

eliarbel commented on qiskit/circuit/random/__init__.py:
```diff
@@ -10,6 +10,41 @@
+Overview
+========
+
+The :mod:qiskit.circuit.random module offers functions that can be used for generating 
```
```suggestion
The :mod:`qiskit.circuit.random` module offers functions that can be used for generating 
```

And this please.

MozammilQ committed:
applied Eli Arbel's suggestions

MozammilQ commented:
After https://github.com/Qiskit/qiskit/pull/12474/commits/53f76280f615a4bf890e5fd32a7097f15a96b702
Roughly speaking there is a performance gain of ~1.6x seen in case of both `random_circuit` and `random_circuit_from_graph`, and their relative speed difference stands roughly the same. Of-course the observation varies with different `seed`.
@eliarbel , applied your suggestions :)

eliarbel committed:
Merge branch 'main' into add_random_circuit_with_graph

MozammilQ committed:
Merge branch 'Qiskit:main' into add_random_circuit_with_graph

MozammilQ commented:
@eliarbel , I was curious enough to update the branch, I hope I didn't mess up anything :)

eliarbel reviewed (approved):
This is ready, thanks @MozammilQ for your contribution!
